okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -661,21 +665,22 @@ private static int delimiterOffset(String input, int pos, int limit, String deli
       return limit;
     }
 
-    private static String username(String input, int pos, int limit) {
-      return input.substring(pos, limit); // TODO: encode
+    private String username(String input, int pos, int limit) {
+      return encode(input, pos, limit, " \"';<=>@[]^`{}|");
     }
 
-    private static String password(String input, int pos, int limit) {
-      return pos < limit ? input.substring(pos, limit) : ""; // TODO: encode
+    private String password(String input, int pos, int limit) {
+      return encode(input, pos, limit, " \"':;<=>@[]\\^`{}|");
     }
 
     private static String host(String input, int pos, int limit) {
       return input.substring(pos, limit); // TODO: encode
     }
 
-    private static int port(String input, int pos, int limit) {
+    private int port(String input, int pos, int limit) {
```

Comment:
I'd experimented with making `utf8Buffer` a field of the builder, rather than a throwaway object used by `encode()`. Changed my mind but didn't put the statics back. I don't feel strongly either way!